File: go/20_valid-parentheses.go
type strStack []string

// Implementation of Stack
func (stack *strStack) push(s string) {
	*stack = append(*stack, s)
}
func (stack *strStack) pop() string {
	result := (*stack)[len(*stack)-1]
	*stack = (*stack)[:(len(*stack) - 1)]
	return result
}
func (stack *strStack) isEmpty() bool {
	if len(*stack) == 0 {
		return true
	}
	return false
}

// Judge whether the stack contains "{", "(" or "["
func contain(s string, dict map[string]string) bool {
	for index := range dict {
		if s == index {
			return true
		}
	}
	return false
}

func isValid(s string) bool {
	dict := map[string]string{"(": ")", "{": "}", "[": "]"}
	var stack strStack = make([]string, 0)

	for _, char := range s {
		if contain(string(char), dict) == true {
			stack.push(string(char))
		} else {
			if stack.isEmpty() == true {
				return false
			}
			tmp := stack.pop()
			if dict[tmp] != string(char) {
				return false
			}
		}
	}
	if stack.isEmpty() == true {
		return true
	}
	return false
}

File: python/20_valid-parentheses.py
class my_stack:
    def __init__(self):
        self.items = []
    def push(self, item):
        self.items.append(item)
    def pop(self):
        return self.items.pop()
    def peek(self):
        return self.items[len(self.items) - 1]
    def is_empty(self):
        return self.items == []
    def size(self):
        return len(self.items)

class Solution:
    def isValid(self, s: str) -> bool:

        stack = my_stack()
        table = {'(': ')', '{': '}', '[': ']'}
        judge = True

        for l in s:
            if l in table:
                stack.push(l)
            else:
                if stack.is_empty():
                    judge = False
                    break
                tmp = stack.pop()
                if table[tmp] == l:
                    continue
                else:
                    judge = False
                    break
        if stack.size() == 0 and judge:
            return True
        else:
            return False